cupy/core/core.pyx

```diff
@@ -1717,51 +1719,177 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
-            pinned_memory._add_to_watch_list(stream.record(), mem)
+
+cdef ndarray _send_object_to_gpu(obj, dtype, str order, Py_ssize_t ndmin):
+    if order is not None and len(order) >= 1 and order[0] in 'KAka':
+        if isinstance(obj, numpy.ndarray) and obj.flags.f_contiguous:
+            order = 'F'
         else:
-            warnings.warn(
-                'Using synchronous transfer as pinned memory ({} bytes) '
-                'could not be allocated. '
-                'This generally occurs because of insufficient host memory. '
-                'The original error was: {}'.format(nbytes, error),
-                util.PerformanceWarning)
-            a.data.copy_from_host(
-                ctypes.c_void_p(a_cpu.__array_interface__['data'][0]), nbytes)
+            order = 'C'
+    a_cpu = numpy.array(obj, dtype=dtype, copy=False, order=order,
+                        ndmin=ndmin)
+    a_dtype = a_cpu.dtype
+    if a_dtype.char not in '?bhilqBHILQefdFD':
+        raise ValueError('Unsupported dtype %s' % a_dtype)
+    cdef vector.vector[Py_ssize_t] a_shape = (<object>a_cpu).shape
+    cdef ndarray a = ndarray(a_shape, dtype=a_dtype, order=order)
+    if a_cpu.ndim == 0:
+        a.fill(a_cpu[()])
+        return a
+    cdef Py_ssize_t nbytes = a.nbytes
+
+    stream = stream_module.get_current_stream()
+    cdef pinned_memory.PinnedMemoryPointer mem = _alloc_pinned_memory(nbytes)
+    if mem is not None:
+        src_cpu = numpy.frombuffer(mem, a_dtype, a_cpu.size)
+        src_cpu[:] = a_cpu.ravel(order)
+        a.data.copy_from_host_async(ctypes.c_void_p(mem.ptr), nbytes)
+        pinned_memory._add_to_watch_list(stream.record(), mem)
+    else:
+        a.data.copy_from_host(
+            ctypes.c_void_p(a_cpu.__array_interface__['data'][0]), nbytes)
 
     return a
 
 
+cdef ndarray _send_numpy_array_list_to_gpu(
+    list arrays, dtype, const vector.vector[Py_ssize_t]& shape, str order,
+        Py_ssize_t ndmin):
+    cdef ndarray a  # allocate it after pinned memory is secured
+    cdef size_t itemcount = internal.prod(shape)
+    cdef size_t nbytes = itemcount * get_dtype(dtype).itemsize
+
+    stream = stream_module.get_current_stream()
+    cdef pinned_memory.PinnedMemoryPointer mem = _alloc_pinned_memory(nbytes)
+    cdef size_t offset, length
+    if mem is not None:
+        # write concatenated arrays to the pinned memory directly
+        src_cpu = numpy.frombuffer(mem, dtype, itemcount).reshape(
+            shape, order=order)
+        numpy.concatenate([e[None] for e in arrays], 0, src_cpu)
+        a = ndarray(shape, dtype=dtype, order=order)
+        a.data.copy_from_host_async(ctypes.c_void_p(mem.ptr), nbytes)
+        pinned_memory._add_to_watch_list(stream.record(), mem)
+    else:
+        # fallback to numpy array and send it to GPU
+        a = _send_object_to_gpu(arrays, dtype, order, ndmin)
+
+    return a
+
+
+cpdef inline _alloc_pinned_memory(Py_ssize_t nbytes):
```

Comment:
Isn't it possible to make this function do the transfer as well, not only allocation?
The warning looks out of place.

Reply:
Now this is used by both `_send_object_to_gpu` and `_send_numpy_array_list_to_gpu` which have different transfer logics. One idea is that move the warning to the above functions. Do you have any ideas?

Reply:
I remembered that it is hard to move to other places because the message takes `e` as an argument.

Reply:
Renamed to `_alloc_async_tranfer_buffer` for clarity.